File: backend/src/main/java/com/centsibility/backend/service/TransactionService.java
package com.centsibility.backend.service;

import com.centsibility.backend.dto.TransactionDTO;
import com.centsibility.backend.model.Transaction;
import com.centsibility.backend.model.User;
import com.centsibility.backend.repository.TransactionRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

/**
 * TransactionService - Business logic for transaction management
 * 
 * TODO: 
 * - Implement full CRUD operations
 * - Add validation logic (amount > 0, valid categories, etc.)
 * - Implement pagination support
 * - Add financial calculations (totals, balance, etc.)
 * - Verify user owns transaction before update/delete
 */
@Service
@Transactional
public class TransactionService {
    
    @Autowired
    private TransactionRepository transactionRepository;
    
    /**
     * Create a new transaction
     * TODO: Implement this method
     * - Validate input data
     * - Set current user as owner
     * - Save to database
     * - Return DTO
     */
    public TransactionDTO createTransaction(TransactionDTO transactionDTO, User user) {
        // TODO: Implement
        throw new UnsupportedOperationException("Not implemented yet");
    }
    
    /**
     * Get all transactions for a user
     * TODO: Implement with pagination
     */
    public List<TransactionDTO> getUserTransactions(User user) {
        // TODO: Implement
        return transactionRepository.findByUser(user).stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }
    
    /**
     * Get transaction by ID
     * TODO: Implement with ownership verification
     */
    public TransactionDTO getTransactionById(Long id, User user) {
        // TODO: Implement
        throw new UnsupportedOperationException("Not implemented yet");
    }
    
    /**
     * Update existing transaction
     * TODO: Implement with ownership verification
     */
    public TransactionDTO updateTransaction(Long id, TransactionDTO transactionDTO, User user) {
        // TODO: Implement
        throw new UnsupportedOperationException("Not implemented yet");
    }
    
    /**
     * Delete transaction
     * TODO: Implement with ownership verification
     */
    public void deleteTransaction(Long id, User user) {
        // TODO: Implement
        throw new UnsupportedOperationException("Not implemented yet");
    }
    
    /**
     * Get transactions by type (INCOME/EXPENSE)
     * TODO: Implement
     */
    public List<TransactionDTO> getTransactionsByType(User user, Transaction.TransactionType type) {
        // TODO: Implement
        throw new UnsupportedOperationException("Not implemented yet");
    }
    
    /**
     * Get transactions within date range
     * TODO: Implement
     */
    public List<TransactionDTO> getTransactionsByDateRange(User user, LocalDateTime start, LocalDateTime end) {
        // TODO: Implement
        throw new UnsupportedOperationException("Not implemented yet");
    }
    
    // Helper method to convert Entity to DTO
    private TransactionDTO convertToDTO(Transaction transaction) {
        return new TransactionDTO(
            transaction.getId(),
            transaction.getAmount(),
            transaction.getType(),
            transaction.getCategory(),
            transaction.getDescription(),
            transaction.getTransactionDate(),
            transaction.getCreatedAt()
        );
    }
    
    // Helper method to convert DTO to Entity
    private Transaction convertToEntity(TransactionDTO dto) {
        Transaction transaction = new Transaction();
        transaction.setAmount(dto.getAmount());
        transaction.setType(dto.getType());
        transaction.setCategory(dto.getCategory());
        transaction.setDescription(dto.getDescription());
        transaction.setTransactionDate(dto.getTransactionDate());
        return transaction;
    }
}

File: web/src/services/transactionService.js
import api from './api';

const transactionService = {
  createTransaction: async (transactionData) => {
    const response = await api.post('/transactions', transactionData);
    return response.data;
  },

  getAllTransactions: async () => {
    const response = await api.get('/transactions');
    return response.data;
  },

  getTransactionById: async (id) => {
    const response = await api.get(`/transactions/${id}`);
    return response.data;
  },

  getTransactionsByDateRange: async (startDate, endDate) => {
    const response = await api.get('/transactions/date-range', {
      params: { startDate, endDate },
    });
    return response.data;
  },

  getTransactionsByType: async (type) => {
    const response = await api.get(`/transactions/type/${type}`);
    return response.data;
  },

  getTransactionsByCategory: async (category) => {
    const response = await api.get(`/transactions/category/${category}`);
    return response.data;
  },

  updateTransaction: async (id, transactionData) => {
    const response = await api.put(`/transactions/${id}`, transactionData);
    return response.data;
  },

  deleteTransaction: async (id) => {
    const response = await api.delete(`/transactions/${id}`);
    return response.data;
  },
};

export default transactionService;
